---
title: "R Notebook"
output: html_notebook
---

Quick check of the first no-genetic data demultiplex sequencing from the pilot.
NPCs

```{r}
library(Seurat)
library(tidyverse)
#library(DoubletFinder)
library(enrichR)
library(clustree)
#library("scClassify")
#library(SingleCellExperiment)
#library("Matrix")


```
Read in the data

```{r}
seu <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/seu_souporcell_ADHD.rds")
colnames(seu@meta.data) # shows you the meta data

unique(seu$orig.ident)
unique(seu$souporcell_assignment)
unique(seu$souporcell_status)
dim(seu)


```

```{r}

table(seu$souporcell_assignment)

```

Look at the sequencing

```{r}

VlnPlot(seu, features = c("nCount_RNA","nFeature_RNA"), pt.size = 0.001)
VlnPlot(seu, features = c("nFeature_RNA"), pt.size = 0.001, y.max = 500)
VlnPlot(seu, features = c("nCount_RNA"), pt.size = 0.001, y.max = 50000)
VlnPlot(seu, features = c("nCount_RNA"), pt.size = 0.001, y.max = 5000)

seu$nFeature_RNA %>% summary
seu$nCount_RNA %>% summary

```

```{r}

table(seu$souporcell_status)
dim(seu)


```




Calculate the percent mitochondrial genes
```{r}
seu <- PercentageFeatureSet(seu, pattern = "^MT-", col.name = "percent.MT")
seu$percent.MT %>% summary

VlnPlot(seu, features = "percent.MT", pt.size = 0.001)
VlnPlot(seu, features = "percent.MT", pt.size = 0.001, y.max = 20)


```
I'll use 10% as a cutoff and check if this removes more of the low read cells. I believe this object is already the CellRanger Filtered object. 

```{r}
#Remove any cells with more than 10% mitochondrial counts
#seu.ft <- subset(seu, percent.MT < 10)
#dim(seu)
#dim(seu.ft)
# removed 656 cells
seu.ft <- subset(seu, percent.MT < 20)
dim(seu)
dim(seu.ft)

# removed 339 cells
VlnPlot(seu.ft, features = c("nCount_RNA","nFeature_RNA","percent.MT"), pt.size = 0.001)


```

```{r}
# have a look at how the counts and RNA are distributed across single, doublet, unassigned

VlnPlot(seu, features = c("nCount_RNA","nFeature_RNA"), group.by = "souporcell_status")
# see most of the unassigned have low counts and features and doublets have higher counts and higher features

VlnPlot(seu, features = c("nCount_RNA"), group.by = "souporcell_status", y.max = 6000)
VlnPlot(seu, features = c("nFeature_RNA"), group.by = "souporcell_status", y.max = 3000)



```
I will filter for higher counts and lower counts


```{r}
# we aren't adding any more filters here

```
clean up

```{r}

seu <- seu.ft
rm(seu.ft)

```


Remove the doublets and the unassigned cells

```{r}
# dim (seu) features: 33538  cells: 17181
Idents(seu) <- "souporcell_status"
levels(seu)
seu <- subset(seu, idents = "singlet")

dim(seu)
# 33538   cells: 14643

```



Now the workflow to make clusters
```{r}
seu <- NormalizeData(seu, normalization.method = "LogNormalize", scale.factor = 10000)
seu <- FindVariableFeatures(seu, selection.method = "vst", nfeatures = 2500)

var  <- VariableFeatures(seu)
top10 <- var[1:10]

p1 <- VariableFeaturePlot(seu) 
p2 <- LabelPoints(plot = p1, points = top10, repel = TRUE)
p2



```
Scale and get PCA

```{r}


#Linear dimensionality reduction
#Choosing the number of PCs can depend on how many cells you have
seu <- ScaleData(seu)
seu <- RunPCA(seu, assay = "RNA", npcs = 30)

PCAPlot(seu)

#Assess how many PCs capture most of the information in the data 
ElbowPlot(seu, ndims = 30)

```

20 PCs are good now make the UMAP

```{r}

seu <- RunUMAP(seu, dims = 1:20, n.neighbors = 121)
DimPlot(seu, group.by = "orig.ident")

```

Find clusters

```{r}

seu <- FindNeighbors(seu, dims = 1:20, k.param = 121)
# the number of clusters is dependent on the resolution a number from 0-2. 
# Higher values make more clusters
# we include 
seu <- FindClusters(seu, resolution = c(0,0.05,0.25,0.6,1,1.5) )
#
# we can visualize which cells are grouped together at different resolutions using clustree

clustree(seu, prefix = "RNA_snn_res.")

# 0.6 looks good to annotate.  Each cluster is splitting apart up to this point and then the cells start merging and changing clusters.

```
Visualize

```{r}
DimPlot(seu, group.by = "RNA_snn_res.0.6", label = TRUE)
DimPlot(seu, group.by = "RNA_snn_res.1", label = TRUE)
# clustering could be improved

```

Look at expression profiles of known cell type markers

```{r}


da_neurons <- c("TH","SLC6A3","SLC18A2","SOX6","NDNF","SNCG","ALDH1A1","CALB1","TACR2","SLC17A6","SLC32A1","OTX2","GRP","LPL","CCK","VIP")
NPC_orStemLike <- c("DCX","NEUROD1","TBR1","PCNA","MKI67","SOX2","NES","PAX6","MASH1")
mature_neurons = c("RBFOX3","SYP","DLG45","VAMP1","VAMP2","TUBB3","SYT1","BSN","HOMER1","SLC17A6")
excitatory_neurons = c("GRIA2","GRIA1","GRIA4","GRIN1","GRIN2B","GRIN2A","GRIN3A","GRIN3","GRIP1","CAMK2A")
inhbitory_neurons = inh = c("GAD1","GAD2", "GAT1","PVALB","GABR2","GABR1","GBRR1","GABRB2","GABRB1","GABRB3","GABRA6","GABRA1","GABRA4","TRAK2")
astrocytes <- c("GFAP","S100B","AQP4","APOE", "SOX9","SLC1A3")
oligodendrocytes <- c("MBP","MOG","OLIG1","OLIG2","SOX10")
opc <- 
radial_glia <- c("PTPRC","AIF1","ADGRE1", "VIM", "TNC","PTPRZ1","FAM107A","HOPX","LIFR",
              "ITGB5","IL6ST","SLC1A3")
epithelial <- c("HES1","HES5","SOX2","SOX10","NES","CDH1","NOTCH1")

microglia <- c("IBA1","P2RY12","P2RY13","TREM119", "GPR34","SIGLECH","TREM2",
               "CX3CR1","FCRLS","OLFML3","HEXB","TGFBR1", "SALL1","MERTK",
               "PROS1")

features_list <- c("MKI67","SOX2","POU5F1","DLX2","PAX6","SOX9","HES1","NES","RBFOX3","MAP2","NCAM1","CD24","GRIA2","GRIN2B","GABBR1","GAD1","GAD2","GABRA1","GABRB2","TH","ALDH1A1","LMX1B","NR4A2","CORIN","CALB1","KCNJ6","CXCR4","ITGA6","SLC1A3","CD44","AQP4","S100B", "PDGFRA","OLIG2","MBP","CLDN11","VIM","VCAM1")

short_list <- c("MKI67","SOX9","HES1","NES","DLX2","RBFOX3","MAP2","TH","CALB1","KCNJ6","SLC1A3","CD44","AQP4","S100B","OLIG2","MBP","VIM")

# musashi-2 is MSI2
NPC_c <- c("FOXG1","SOX2","TUBB3","NES","MSI2","DKK1","DKK3","CTNNB1","LRP6")


```

View on UMAP

```{r}

Idents(seu) <- "RNA_snn_res.0.6"

for (i in NPC_c) {
  print(FeaturePlot(seu, features = i, min.cutoff = 'q1', max.cutoff = 'q97', label = TRUE))
}

```

```{r}
Idents(seu) <- "RNA_snn_res.0.6"

for (i in astrocytes) {
  print(FeaturePlot(seu, features = i, min.cutoff = 'q1', max.cutoff = 'q97', label = TRUE))
}

```

```{r}

Idents(seu) <- "RNA_snn_res.0.6"

for (i in radial_glia) {
  print(FeaturePlot(seu, features = i, min.cutoff = 'q1', max.cutoff = 'q97', label = TRUE))
}

```

```{r}
Idents(seu) <- "RNA_snn_res.0.6"

mature_neuronsB = c("RBFOX3","SYP","VAMP1","VAMP2","TUBB3","SYT1","BSN","HOMER1","SLC17A6")
for (i in mature_neuronsB) {
  print(FeaturePlot(seu, features = i, min.cutoff = 'q1', max.cutoff = 'q97', label = TRUE))
}



```


Dot plots and heatmaps

```{r}
DotPlot(seu, features = radial_glia)+ RotatedAxis()
DotPlot(seu, features = NPC_orStemLike) + RotatedAxis()
DotPlot(seu, features = astrocytes) + RotatedAxis()
DotPlot(seu, features = NPC_c) + RotatedAxis()

```

Find cluster markers

```{r}
Idents(seu) <- "RNA_snn_res.0.6"
ClusterMarkers <- FindAllMarkers(seu, only.pos = TRUE)

```
Have a look at the top cluster markers
```{r}
head(ClusterMarkers)
top5 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=5, wt =avg_log2FC)

top2 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=2, wt =avg_log2FC)
DoHeatmap(seu, features = top5$gene, size = 3, angle = 90, group.by = "RNA_snn_res.0.6")

DoHeatmap(seu, features = top2$gene, size = 3, angle = 90, group.by = "RNA_snn_res.0.6")
```

Now use EnrichR to check cell type libraries

```{r}
setEnrichrSite("Enrichr") # Human genes
# list of all the databases
# get the possible libraries
dbs <- listEnrichrDbs()

# this will list the possible libraries
dbs

# select libraries with cell types
db <- c('CellMarker_Augmented_2021','Azimuth_Cell_Types_2021')

# function for a quick look
checkCelltypes <- function(cluster_num = 0){
  clusterX <- ClusterMarkers %>% filter(cluster == cluster_num & avg_log2FC > 0.25)
  genes <- clusterX$gene
  # the cell type libraries
  # get the results for each library
  clusterX.cell <- enrichr(genes, databases = db)
  # visualize the results
print(plotEnrich(clusterX.cell[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'CellMarker_Augmented_2021'))
print(plotEnrich(clusterX.cell[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'Azimuth_Cell_Types_2021'))

}


```

Check each cluster
```{r}
cluster0 <- checkCelltypes(cluster_num = 0)

```
```{r}
cluster1 <- checkCelltypes(cluster_num = 1)
```

```{r}
cluster2 <- checkCelltypes(cluster_num = 2)
```

```{r}
cluster3 <- checkCelltypes(cluster_num = 3)
```
```{r}
cluster4 <- checkCelltypes(cluster_num = 4)
```
```{r}
cluster5 <- checkCelltypes(cluster_num = 5)
```
```{r}
cluster6 <- checkCelltypes(cluster_num = 6)
```
```{r}
cluster7 <- checkCelltypes(cluster_num = 7)
```
```{r}
cluster8 <- checkCelltypes(cluster_num = 8)
```
```{r}
cluster9 <- checkCelltypes(cluster_num = 9)
```

Add cell type annotations

```{r}
# we need to set the identity to rename
Idents(seu) <- "RNA_snn_res.0.6"

# we need to make a vector of the cell type in the same order - in the cluster order

cell_types <- c("NPC-pluri", "NPC","NPC",
                 "glia","glia","Neuron",
                 "glia-APOE","NPC-mixed","NPC",
                 "NPC-TNC")
names(cell_types) <- levels(seu)

seu <- RenameIdents(seu, cell_types)
seu <- AddMetaData(object=seu, metadata=Idents(seu), col.name = "CellTypes")

DimPlot(seu, label = TRUE)

Idents(seu) <- "RNA_snn_res.0.6"
cell_types <- c("NPC-pluri", "NPC1","NPC2",
                 "glia1","glia2","Neuron",
                 "glia3","NPC-mixed","NPC-div",
                 "NPC-TNC")
names(cell_types) <- levels(seu)
seu <- RenameIdents(seu, cell_types)
seu <- AddMetaData(object=seu, metadata=Idents(seu), col.name = "Celltypes1")

DimPlot(seu, label = TRUE)


```

See cell counts for each line and cell type

```{r}
cdf <- as.data.frame(table(seu$souporcell_assignment,seu$CellTypes))

df_wide <- cdf %>%
  pivot_wider(names_from = Var2, values_from = Freq)

df_wide

write.csv(df_wide, "/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/soup_ADHD_cellcountslinetypeJune28.csv")

```


```{r}
VlnPlot(seu, features = "nFeature_RNA", group.by = "souporcell_assignment")

Idents(seu)<- "CellTypes"
DimPlot(seu, group.by = "souporcell_assignment", label = TRUE)

```

We might need to separate and harmonize each cell type.

```{r}
# make a list of seurat objects by our cell type variable
sublist <- SplitObject(seu, split.by = "souporcell_assignment")
# normalize and find variable features
for (i in 1:length(sublist)){
  sublist[[i]] <- NormalizeData(sublist[[i]], verbose = FALSE)
  sublist[[i]] <- FindVariableFeatures(sublist[[i]], selection.method = "vst")
}



```

```{r}
# Create an empty Seurat object to store the integrated data
# Take the first Seurat object from the list as the starting point
integrated_seurat <- subset(sublist[[1]])


# Iterate over the list of Seurat objects
for (i in 1:length(sublist)) {
  # Rename the 'orig.ident' metadata inside the seurat object to match the object name in the list
  sublist[[i]]$orig.ident <- names(sublist)[i]

}

sample.list <- sublist
for (i in 1:length(sample.list)) {
  # Normalize and scale the data
  sample.list[[i]] <- NormalizeData(sample.list[[i]], verbose = FALSE)
  sample.list[[i]] <- ScaleData(sample.list[[i]], verbose = FALSE)
  # Find variable features
  sample.list[[i]] <- FindVariableFeatures(sample.list[[i]], selection.method = "vst")
  # Get the variable features
  variable_features <- VariableFeatures(sample.list[[i]])
  # Run PCA with the variable features
  sample.list[[i]] <- RunPCA(sample.list[[i]], verbose = FALSE, npcs = 30, features = variable_features)
}

int.anchors <- FindIntegrationAnchors(object.list = sample.list, dims = 1:30, reduction = "rpca")
integrated_seurat <- IntegrateData(anchorset = int.anchors,  dims = 1:30, k.weight = 72)
# the k value is the problem
# must set the k weight to the lowest cell count


```
Save so far

```{r}
saveRDS(seu, "/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/seu_ann_June28.RDS")
saveRDS(integrated_seurat, "/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/Integr_seu_ann_June28.RDS")

seu <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/Integr_seu_ann_June28.RDS")


```

Check the object

```{r}
#library(Seurat)
colnames(seu@meta.data)

```


```{r}
DimPlot(seu, group.by = "CellTypes")
DimPlot(seu, group.by = "CellTypes1")
DimPlot(seu, group.by = "Celltypes1")
DimPlot(seu, group.by = "CellTypesInt")
DimPlot(seu, group.by = "CellTypesInt2")

```





```{r}
DimPlot(seu, group.by = "souporcell_assignment", split.by = "CellTypes")

DimPlot(seu)

```





```{r}

DefaultAssay(integrated_seurat) <- "integrated"
integrated_seurat <- ScaleData(integrated_seurat, verbose = FALSE)
# only the integrated features will be the pca input

integrated_seurat <- RunPCA(integrated_seurat, npcs = 20, verbose = FALSE)
integrated_seurat <- RunUMAP(integrated_seurat, reduction = "pca", dims = 1:20, n.neighbors = 81)

DimPlot(integrated_seurat, group.by = "souporcell_assignment")

```
```{r}
DimPlot(integrated_seurat, group.by = "CellTypes")


```
```{r}

DimPlot(integrated_seurat, group.by = "souporcell_assignment", split.by = "CellTypes")


```


```{r}

integrated_seurat <- FindNeighbors(integrated_seurat, dims = 1:20, k.param = 81)
integrated_seurat <- FindClusters(integrated_seurat, resolution = c(0.6,1) )

```
```{r}
DimPlot(integrated_seurat, group.by = "orig.ident")
#much better integration
```
```{r}
#integrated_seurat$integrated_snn_res.0.6
DimPlot(integrated_seurat, group.by = "integrated_snn_res.0.6", label=TRUE)
DimPlot(integrated_seurat, group.by = "integrated_snn_res.1", label = TRUE)

```

Now cells per cluster

```{r}
table(integrated_seurat$integrated_snn_res.0.6, integrated_seurat$orig.ident)

table(integrated_seurat$integrated_snn_res.1, integrated_seurat$orig.ident)
```

Which cells match to which subject?

From Michaels correlation:
```{r}
code.tb <- read.table(file = "/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/Genotype_ID_key.txt", 
header = TRUE, sep = "\t")

class(code.tb)


```
```{r}

# add the assigned lines into the seurat object
# Assuming your data frame is named 'lookup_df' and has columns 'Genotype_ID' and 'Cluster_ID'
lookup_table <- code.tb[, c("Cluster_ID", "Genotype_ID")]
# add the a new metadata slot
seu$Genotype_ID <- lookup_table$Genotype_ID[match(seu$souporcell_assignment, lookup_table$Cluster_ID)]


seu$Genotype_ID <- ifelse(
  seu$souporcell_assignment %in% lookup_table$Cluster_ID,
  lookup_table$Genotype_ID[match(seu$souporcell_assignment, lookup_table$Cluster_ID)],
  "unassigned"
)



unique(seu$Genotype_ID)

```
Add genotype to intergrated object

```{r}
lookup_table <- code.tb[, c("Cluster_ID", "Genotype_ID")]
# add the a new metadata slot
integrated_seurat$Genotype_ID <- lookup_table$Genotype_ID[match(integrated_seurat$souporcell_assignment, lookup_table$Cluster_ID)]


integrated_seurat$Genotype_ID <- ifelse(
  integrated_seurat$souporcell_assignment %in% lookup_table$Cluster_ID,
  lookup_table$Genotype_ID[match(integrated_seurat$souporcell_assignment, lookup_table$Cluster_ID)],
  "unassigned"
)



unique(integrated_seurat$Genotype_ID)

```

Integrated object annotations

```{r}
table(integrated_seurat$CellTypes, integrated_seurat$integrated_snn_res.0.6)

```

```{r}

Idents(integrated_seurat) <- "integrated_snn_res.0.6"
ClusterMarkers <- FindAllMarkers(integrated_seurat, only.pos = TRUE)  


```



```{r}

checkCelltypes(cluster_num = 12)

```

Add annotation

```{r}

# we need to set the identity to rename
Idents(integrated_seurat) <- "integrated_snn_res.0.6"

# we need to make a vector of the cell type in the same order - in the cluster order

cell_types <- c("RG", "NPC-stem","glia-stem",
                 "NPC-div1","Neuron","NPC-stem2","NPC-div2",
                 "NPC-mixed","glia","astro","NPC-div3",
                 "NPC-pluri","glia-RG")
names(cell_types) <- levels(integrated_seurat)

integrated_seurat <- RenameIdents(integrated_seurat, cell_types)
integrated_seurat <- AddMetaData(object=integrated_seurat, metadata=Idents(integrated_seurat), col.name = "CellTypesInt")

DimPlot(integrated_seurat, label = TRUE, label.size = 2)

Idents(integrated_seurat) <- "integrated_snn_res.0.6"

# we need to make a vector of the cell type in the same order - in the cluster order

cell_types <- c("RG", "NPC-stem","glia",
                 "NPC-div","Neuron","NPC-stem","NPC-div",
                 "NPC-mixed","glia","astro","NPC-mixed",
                 "NPC-stem","glia2")
names(cell_types) <- levels(integrated_seurat)

integrated_seurat <- RenameIdents(integrated_seurat, cell_types)
integrated_seurat <- AddMetaData(object=integrated_seurat, metadata=Idents(integrated_seurat), col.name = "CellTypesInt2")

DimPlot(integrated_seurat, label = TRUE, label.size = 2)

```


Now make a table of the assigned cell counts

```{r}

seu <- integrated_seurat

gene.count <- as.data.frame(table(seu$Genotype_ID))

group.gene <- as.data.frame(table(seu$souporcell_assignment, seu$Genotype_ID))

soup.count <- as.data.frame(table(seu$souporcell_assignment))

# Assuming your long table is named 'long_table'
wide_table <- group.gene %>%
  pivot_wider(names_from = Var1, values_from = Freq)

wide_table
write.csv(wide_table,"/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/SoupoGeneIDJune28.csv")

```


Read in annotated object saved June 28

```{r}
seu <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/Integr_seu_ann_June28.RDS")

```

Check if it's correct

```{r}
DimPlot(seu)

```

Add shorter ID names / cell line names

```{r}
Idents(seu) <- "Genotype_ID"
levels(seu)

line.clone <- c("MR010i3","KO15i1","KO8i13","KO11i10",
              "unassigned","MR014i27","MR023i17","MR01i3","NR2i21","K05i12")

names(line.clone) <- levels(seu)

seu <- RenameIdents(seu, line.clone)
seu <- AddMetaData(object=seu, metadata=Idents(seu), col.name = "line.clone")

DimPlot(seu, label = FALSE)



```


Add meta data - diagnosis

```{r}

#seu$souporcell_assignment
#seu$Genotype_ID
Idents(seu) <- "Genotype_ID"
levels(seu)

diagnose <- c("ADHD","Control","Control","Control",
              "Unassigned","ADHD","ADHD","ADHD","ADHD-non","Control")

names(diagnose) <- levels(seu)

seu <- RenameIdents(seu, diagnose)
seu <- AddMetaData(object=seu, metadata=Idents(seu), col.name = "Diagnosis")

DimPlot(seu, label = FALSE)

```

Differential gene expression by cell type

```{r}
colnames(seu@meta.data)
# subset the cell type we are interested
Idents(seu) <- "CellTypesInt2"
levels(seu)

```



```{r}
# get just the NPC-div cells
DefaultAssay(seu) <- "RNA"
NPC_div <- subset(seu, idents = "NPC-div")
levels(NPC_div)

```

```{r}
# DGE
NPC_div <- ScaleData(NPC_div)

Idents(NPC_div) <- "Diagnosis"
NPC_div_DGE <- FindMarkers(NPC_div, ident.1 = "ADHD",
                           ident.2 = "Control", logfc.threshold = 0.01)

write.csv(NPC_div_DGE, "/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/NPC_div_DGE.csv")

NPC_div <- read.csv("/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/NPC_div_DGE.csv")

```

Make volcano plot

```{r, fig.height= 2.5}
#library(EnhancedVolcano)

#head(NPC_div)



EnhancedVolcano(NPC_div,
    lab = NPC_div$X,
    xlim = c(-0.25,0.25),
    x = 'avg_log2FC',
    y = 'p_val_adj',
      pCutoff = 10e-32,
    FCcutoff = 0.025,
    pointSize = 3.0,
    labSize = 6.0)




```

Heat map of some interesting genes

```{r}
ft <- c("WNT4","ARX","CA4","GDPD2","PHOX2B","CYTOR","GSX1","PLP1")
DoHeatmap(seu, features = ft, group.by = 
          "Diagnosis")

DoHeatmap(seu, features = ft, group.by = 
          "line.clone")

```


```{r}
ft <- c("WNT4","ARX","CA4","GDPD2","PHOX2B","CYTOR","GSX1","PLP1")
DotPlot(seu, features = ft, group.by = 
          "Diagnosis") +RotatedAxis()

DotPlot(seu, features = ft, group.by = 
          "line.clone")+ RotatedAxis()

```



Double group heat map function



see double variable heatmap

```{r}

#function not working right now
  


```





Check some expression

```{r}

DoHeatmap(NPC_div, group.by = "Diagnosis", features = c("XIST","KRT18","CRABP1",
                                                        "FRZB","A2M","CHCHD2"))

DotPlot(NPC_div, group.by = "Diagnosis", features = c("XIST","KRT18","CRABP1",
                                                        "FRZB","A2M","CHCHD2"))+RotatedAxis()

```

NPC-stem

```{r}

Idents(seu) <- "CellTypesInt2"

# get just the NPC-stem cells
NPC_stem <- subset(seu, idents = "NPC-stem")
levels(NPC_stem)

# DGE

Idents(NPC_stem) <- "Diagnosis"
NPC_stem_DGE <- FindMarkers(NPC_stem, ident.1 = "ADHD",
                           ident.2 = "Control", logfc.threshold = 0.01)

write.csv(NPC_stem_DGE, "/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/NPC_stem_DGE.csv")


NPC_stem <- read.csv("/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/NPC_stem_DGE.csv")


```

Volcono NPC stem DGE

```{r, fig.height= 2.5}


EnhancedVolcano(NPC_stem,
    lab = NPC_stem$X,
    xlim = c(-0.25,0.25),
    x = 'avg_log2FC',
    y = 'p_val_adj',
      pCutoff = 10e-32,
    FCcutoff = 0.05,
    pointSize = 3.0,
    labSize = 6.0)

```

Look at some genes in the NPC-stem DGE

```{r}
ft <- c("IFI16","AHNAK","PLA2G16","OLFML3","PDLIM3","CSRP1","MX1","ACKR3"
)
DotPlot(seu, features = ft, group.by = 
          "Diagnosis") +RotatedAxis()

DotPlot(seu, features = ft, group.by = 
          "line.clone")+ RotatedAxis()



```





```{r}

dge <- c("CA4","STMN2","KRT18","HIST1H1A","S100A10","IFI27","CLSPN","IFI6","HELLS", "ACTC1")

DoHeatmap(NPC_div, group.by = "Diagnosis", features = dge)

DotPlot(NPC_div, group.by = "Diagnosis", features = dge)+RotatedAxis()

```

```{r}
Idents(seu) <- "CellTypesInt2"

# get just the neuron cells
neurons <- subset(seu, idents = "Neuron")


# DGE

Idents(neurons) <- "Diagnosis"
neurons_DGE <- FindMarkers(neurons, ident.1 = "ADHD",
                           ident.2 = "Control", logfc.threshold = 0.01)

write.csv(neurons_DGE, "/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/Neurons_DGE.csv")


neurons <- read.csv("/Users/rhalenathomas/Documents/Data/scRNAseq/ADHD_ZNZ_Mcgill/Neurons_DGE.csv")


```


Volcano plot Neurons

```{r, fig.height= 2.5}

EnhancedVolcano(neurons,
    lab = neurons$X,
    xlim = c(-0.25,0.25),
    x = 'avg_log2FC',
    y = 'p_val_adj',
      pCutoff = 10e-32,
    FCcutoff = 0.075,
    pointSize = 3.0,
    labSize = 6.0)

```
Check some genes with the dotplot

```{r}
ft <- c("CRYBB1","ASS1","PDLIM1","GDF15","TMEM176A","LHX1","TEM176B","NHLH2")
DotPlot(seu, features = ft, group.by = 
          "Diagnosis") +RotatedAxis()

DotPlot(seu, features = ft, group.by = 
          "line.clone")+ RotatedAxis()



```



Add some more phenotype information

```{r}
Idents(seu) <- "line.clone"
levels(seu)

sex <- c("M","M","F","M",
              "unassigned-both","M","M","M","M","M")

names(sex) <- levels(seu)

seu <- RenameIdents(seu, sex)
seu <- AddMetaData(object=seu, metadata=Idents(seu), col.name = "sex")

DimPlot(seu, label = FALSE)


```


Balancing the cell numbers more between lines

```{r}
table(seu$line.clone)
```
```{r}
Idents(seu) <- "line.clone"
min.seu <- subset(seu, downsample = 376)
table(min.seu$line.clone)
table(min.seu$Diagnosis)

```

Try differential gene expression 

```{r}
unique(min.seu$CellTypesInt2)
Idents(min.seu) <- "CellTypesInt2"
seu.npc.stem <- subset(min.seu, idents = "NPC-stem")

Idents(seu.npc.stem) <- "Diagnosis"
DGE2_downsample_NPC_stem <- FindMarkers(seu.npc.stem, ident.1 = "ADHD",
                           ident.2 = "Control", logfc.threshold = 0.025)

```


```{r, fig.height= 3}
# plot with all cells included before down sampling
# LFC cutoff of 0.01

EnhancedVolcano(NPC_stem,
    lab = NPC_stem$X,
    xlim = c(-0.5,0.5),
    x = 'avg_log2FC',
    y = 'p_val_adj',
      pCutoff = 2,
    FCcutoff = 0.1,
    pointSize = 3.0,
    labSize = 6.0)

# with downsampling and cutoff for 0.025

EnhancedVolcano(DGE2_downsample_NPC_stem,
    lab = rownames(DGE2_downsample_NPC_stem),
    xlim = c(-0.5,0.5),
    x = 'avg_log2FC',
    y = 'p_val_adj',
      pCutoff = 2,
    FCcutoff = 0.1,
    pointSize = 3.0,
    labSize = 6.0)

```

```{r}
unique(seu$CellTypesInt2)
unique(seu$CellTypesInt)
unique(seu$CellTypes)

```

Check NPC-mix without downsampling

```{r}


Idents(seu) <- "CellTypesInt2"
npc.mix <- subset(seu, idents = "NPC-mixed")

Idents(npc.mix) <- "Diagnosis"
NPC_mix_DGE <- FindMarkers(npc.mix, ident.1 = "ADHD",
                           ident.2 = "Control", logfc.threshold = 0.025)

```


Volxano plot

```{r, fig.height= 3}

EnhancedVolcano(NPC_mix_DGE,
    lab = rownames(NPC_mix_DGE),
    xlim = c(-1,1),
    x = 'avg_log2FC',
    y = 'p_val',
    #pCutoff = 10,
    FCcutoff = 0.2,
    pointSize = 3.0,
    labSize = 6.0)


```

```{r}
head(NPC_mix_DGE)
```



Do a GO analysis 

```{r}
library(enrichR)
setEnrichrSite("Enrichr") # Human genes
# list of all the databases
# get the possible libraries
dbs <- listEnrichrDbs()

# this will list the possible libraries
dbs

# select libraries with cell types
db <- c('KEGG_2019_Human','GWAS_Catalog_2019',"GO_Biological_Process_2023",
        "GO_Cellular_Component_2023","GO_Molecular_Function_2023")

dataframe <- NPC_mix_DGE
library(dplyr)

# add genes column to DGE it is X if saved to csv
NPC_mix_DGE$X <- rownames(NPC_mix_DGE)

# function for a quick look
getGSA <- function(dataframe, up_or_down = "up", LFCthresh = 0.01,
                          pval_thresh = 0.05){
  if(up_or_down == "up"){
      genelist <- dataframe %>% filter(avg_log2FC >= LFCthresh & p_val_adj < pval_thresh)
  }else {
      genelist <- dataframe %>% filter(avg_log2FC <= LFCthresh & p_val_adj < pval_thresh)
  }
genes <- genelist$X
  # the cell type libraries
  # get the results for each library
  results <- enrichr(genes, databases = db)
  # visualize the results
  
print(plotEnrich(results[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'KEGG'))
print(plotEnrich(results[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'GWAS'))
print(plotEnrich(results[[3]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'GObio'))
print(plotEnrich(results[[4]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'GOcell'))
print(plotEnrich(results[[5]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'GOmol'))
return(results)
}

GONPCmix <- getGSA(NPC_mix_DGE, up_or_down = "up",LFCthresh = 0.02,pval_thresh = 0.05)

resDF <- GONPCmix

plotEnrich(resDF[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'KEGG')


dfGO <- GONPCmix[[1]]
write.csv(dfGO, "KEGGanalysisNPCMixDGEup.csv")

GWAS <- GONPCmix[[2]]
write.csv(GWAS, "GWASNPCMixDGEup.csv")

dfGO <- GONPCmix[[3]]
write.csv(dfGO, "GObioNPCMixDGEup.csv")

dfGO <- GONPCmix[[4]]
write.csv(dfGO, "GOcellNPCMixDGEup.csv")

dfGO <- GONPCmix[[5]]
write.csv(dfGO, "GOmoleNPCMixDGEup.csv")


```

Look at the down regulated (in ADHD NPC-mix)


```{r}

GONPCmix <- getGSA(NPC_mix_DGE, up_or_down = "down",LFCthresh = -0.02,pval_thresh = 0.05)


```




Try pseudobulk

 
```{r}
unique(seu$CellTypesInt2)

Idents(seu) <- "CellTypesInt2"
levels(seu)



groupcells <- c("RG","NPC","glia","NPC","Neuron","NPC","astro","glia")
names(groupcells) <- levels(seu)

seu <- RenameIdents(seu, groupcells)
seu <- AddMetaData(object=seu, metadata=Idents(seu), col.name = "CelltypeMain")

DimPlot(seu, label = TRUE)


```


```{r}

Idents(seu) <- "CelltypeMain"
npc<- subset(seu, idents = "NPC")

Idents(npc) <- "Diagnosis"
NPC_mix_DGE <- FindMarkers(npc, ident.1 = "ADHD",
                           ident.2 = "Control", logfc.threshold = 0.025)
dim(NPC_mix_DGE)


```

Volcano plot

```{r, fig.height=3}


EnhancedVolcano(NPC_mix_DGE,
    lab = rownames(NPC_mix_DGE),
    xlim = c(-0.5,0.5),
    x = 'avg_log2FC',
    y = 'p_val',
    #pCutoff = 10,
    FCcutoff = 0.1,
    pointSize = 3.0,
    labSize = 6.0)

```

We think it's better to keep the sub groups of NPC separate

```{r}
Idents(seu) <- "CelltypeMain"
rg<- subset(seu, idents = "RG")

Idents(rg) <- "Diagnosis"
RG_DGE <- FindMarkers(rg, ident.1 = "ADHD",
                           ident.2 = "Control", logfc.threshold = 0.025)
dim(RG_DGE)


```

```{r, fig.height=3}

EnhancedVolcano(RG_DGE,
    lab = rownames(RG_DGE),
    xlim = c(-0.5,0.5),
    x = 'avg_log2FC',
    y = 'p_val',
    pCutoff = 0.05,
    FCcutoff = 0.1,
    pointSize = 3.0,
    labSize = 6.0)

```



Try pseudobulk
https://hbctraining.github.io/scRNA-seq_online/lessons/pseudobulk_DESeq2_scrnaseq.html

```{r}
# filter to have just the things we want (identified subjects)

Idents(seu) <- "Diagnosis"
seu2 <- subset(seu, idents = c("ADHD","Control"))
table(seu2$Diagnosis)

```


```{r}
# make sce

sce <- as.SingleCellExperiment(seu2)
counts(sce) <- seu2@assays$RNA@counts



```

Prepare single cell data

```{r}
#sce@colData$CellTypesInt2

cluster_names <- levels(sce@colData$CellTypesInt2)
cluster_names
length(cluster_names)

line_names <- levels(sce@colData$line.clone)
line_names
length(line_names)


```

```{r}

library(SingleCellExperiment)

groups <- colData(sce)[,c("CellTypesInt2", "line.clone")]
head(groups)
# Aggregate across cluster-sample groups
# transposing row/columns to have cell_ids as row names matching those of groups

# install from github
library(devtools)
install_github("cvarrichio/Matrix.utils")

library(Matrix.utils)


aggr_counts <- aggregate.Matrix(t(sce@assays@data@listData$logcounts), 
                                groupings = groups, fun = "sum") 

# using log counts for now

# Explore output matrix
class(aggr_counts)
dim(aggr_counts)
aggr_counts[1:6, 1:6]


```

```{r}

# Transpose aggregated matrix to have genes as rows and samples as columns
aggr_counts <- t(aggr_counts)
aggr_counts[1:6, 1:6]

```

```{r}
library(data.table)
# Understanding tstrsplit()

## Exploring structure of function output (list)
tstrsplit(colnames(aggr_counts), "_") %>% str()

## Comparing the first 10 elements of our input and output strings
head(colnames(aggr_counts), n = 10)
head(tstrsplit(colnames(aggr_counts), "_")[[1]], n = 10)

```

Look at a cluster

```{r}

# Using which() to look up tstrsplit() output
npc_idx <- which(tstrsplit(colnames(aggr_counts), "_")[[1]] == "NPC-mixed")
npc_idx

colnames(aggr_counts)[npc_idx]
aggr_counts[1:10, npc_idx]
```

Make a list with agg for each cluster

```{r}

# As a reminder, we stored our cell types in a vector called cluster_names
cluster_names


# Loop over all cell types to extract corresponding counts, and store information in a list

## Initiate empty list
counts_ls <- list()

for (i in 1:length(cluster_names)) {

  ## Extract indexes of columns in the global matrix that match a given cluster
  column_idx <- which(tstrsplit(colnames(aggr_counts), "_")[[1]] == cluster_names[i])
  
  ## Store corresponding sub-matrix as one element of a list
  counts_ls[[i]] <- aggr_counts[, column_idx]
  names(counts_ls)[i] <- cluster_names[i]

}

# Explore the different components of the list
str(counts_ls)
```

Make a meta data structure

```{r}

# Reminder: explore structure of metadata
head(colData(sce))

# Extract sample-level variables
metadata <- colData(sce) %>% 
  as.data.frame() %>% 
  dplyr::select(CellTypesInt2,line.clone, Diagnosis, sex)

dim(metadata)
head(metadata)

# Exclude duplicated rows
metadata <- metadata[!duplicated(metadata), ]

dim(metadata)
head(metadata)
```
```{r}
# Number of cells per sample and cluster
t <- table(colData(sce)$line.clone,
           colData(sce)$CellTypeInt2)
#t[1:6, 1:6]



```

```{r}
# Creating metadata list

## Initiate empty list
metadata_ls <- list()

for (i in 1:length(counts_ls)) {
  
    ## Initiate a data frame for cluster i with one row per sample (matching column names in the counts matrix)
    df <- data.frame(cluster_sample_id = colnames(counts_ls[[i]]))
    
    ## Use tstrsplit() to separate cluster (cell type) and sample IDs
    df$cluster_id <- tstrsplit(df$cluster_sample_id, "_")[[1]]
    df$sample_id  <- tstrsplit(df$cluster_sample_id, "_")[[2]]
    
    
    ## Retrieve cell count information for this cluster from global cell count table
    idx <- which(colnames(t) == unique(df$cluster_id))
    cell_counts <- t[, idx]
    
    ## Remove samples with zero cell contributing to the cluster
    cell_counts <- cell_counts[cell_counts > 0]
    
    ## Match order of cell_counts and sample_ids
    sample_order <- match(df$sample_id, names(cell_counts))
    cell_counts <- cell_counts[sample_order]
    
    ## Append cell_counts to data frame
    df$cell_count <- cell_counts
    
    
    ## Join data frame (capturing metadata specific to cluster) to generic metadata
    df <- plyr::join(df, metadata, 
                     by = intersect(names(df), names(metadata)))
    
    ## Update rownames of metadata to match colnames of count matrix, as needed later for DE
    rownames(df) <- df$cluster_sample_id
    
    ## Store complete metadata for cluster i in list
    metadata_ls[[i]] <- df
    names(metadata_ls)[i] <- unique(df$cluster_id)

}

# Explore the different components of the list
str(metadata_ls)

```
DGEseq2

```{r}
idx <- which(names(counts_ls) == "NPC-mixed")
cluster_counts <- counts_ls[[idx]]
cluster_metadata <- metadata_ls[[idx]]


```

checking
```{r}

# Check contents of extracted objects
cluster_counts[1:6, 1:6]
head(cluster_metadata)

# Check matching of matrix columns and metadata rows
all(colnames(cluster_counts) == rownames(cluster_metadata))
```

Try directly in seurat

```{r}

agg.npc <- AggregateExpression(
  npc.mix,
  return.seurat = FALSE,
  group.by = "line.clone",
  add.ident = NULL,
  slot = "data",
  verbose = TRUE
)

#no good

```


Try another website

```{r}

```




